packages/pathfinder/src/pathfinder.ts

```diff
@@ -0,0 +1,400 @@
+                pathInfo.from.code == sourceAsset.code;
+        });
+
+        if (!bestPath) {
+            throw new Error("[getPath]: stellar account not configured correctly");
+        }
+
+        return bestPath;
+    }
+
+    /***************************************************************************
+     * Cache Logic
+     ***************************************************************************/
+
+    /**
+     * Queries the cache and returns a promise for the path information; if no path information
+     * exists, the promise resolves to null.
+     */
+    private async loadPathFromCache(
+        sourceAsset: stellarAsset,
+        destinationAsset: stellarAsset,
+        threshold: number,
+    ): Promise<IPathInfo | null> {
+        // todo provide the retrieval logic
+        const key = `${destinationAsset.code}_${destinationAsset.issuerPublicKey}`;
+        const TEMP_DICT = {};
+        // return new Promise((res: any, rej: any ) => {
+        //      return this.cache.lookup(key)
+        //     Promise.resolve(null);
+        // });
+        return Promise.resolve(null);
+    }
+
+    /**
+     * Saves the path info to the cache for each path in bestPaths
+     *
+     * Note: no need to make this a promise because we will never check up on it
+     */
+    private savePathsToCache(
+        destinationAsset: stellarAsset,
+        bucket: number,
+        bestPaths: IPathInfo[],
+    ) {
+        const keyPrefix = `${destinationAsset.code}_${destinationAsset.issuerPublicKey}-${bucket}-`;
+
+        bestPaths.forEach((pathInfo) => {
+            // todo: need to be able to XLM which doesnt have all asset props
+            const key = `${keyPrefix}${pathInfo.from.code}-${pathInfo.from.issuerPublicKey}`;
+            const value = {
+                exchangeRate: pathInfo.exchangeRate,
+                path: pathInfo.path,
+            };
+            // todo provide the saving mechanism
+            // this.cache.savePaths(key, value)
+        });
+    }
+
+    /**
+     * Computes the relevant bucket for the given destination asset and amount. This is
+     * required because the paths between assets might be different at different amounts; however,
+     * to enable caching we must quantize the destination amounts to constrain the maximum cache size
+     * (e.g., caching every real numbered destination amount that we receive would result in way too
+     * many cache entries and cache misses)
+     *
+     * For example, consider needing to convert eth to btc. The path for 1 btc might be
+     *
+     * eth - btc
+     *
+     * but for larger amounts like 100 btc it might be more complicated
+     *
+     * eth - intermediate 1 - intermediate 2 - intermediate 3 - btc
+     *
+     * //Todo need to decide and implement the quantization method
+     */
+
+    private getBucket(
+        destinationAsset: stellarAsset,
+        destinationAmount: number,
+    ): number {
+        return 10000;
```

Comment:
The bucket is related to calculating the quantized transaction level. If the above description is not clear enough or you are unsure about how to use this, please call me.